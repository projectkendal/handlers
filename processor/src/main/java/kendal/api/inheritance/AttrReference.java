package kendal.api.inheritance;

/**
 * Java does not allow using parameters of annotation to define other parameters of the same annotation.
 * But sometimes we want to, so here we introduce annotation that will be replaced with value of another parameter.
 * It can be used only inside value expression for {@link Attribute}.
 * Some expressions using @AttrReference will cause the compiler to break compilation before annotation processing.
 * In such case @AttrReference will never work. This feature is currently considered a proof of concept showing that
 * it is possible to access other attributes of annotation to define value of attribute.
 */
public @interface AttrReference {

    String value();
}
