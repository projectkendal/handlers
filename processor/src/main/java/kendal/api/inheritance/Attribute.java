package kendal.api.inheritance;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * This annotation will add named parameter to all usages of annotated annotation.
 * Use {@link List} to add multiple parameters.
 * Parameters will be added to usages of annotation before handlers are called.
 *
 * Use {@link AttrReference} as placeholder for values of any other attributes.
 */
@Target(ElementType.ANNOTATION_TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Attribute {

    String name();

    //commented out, but we process it
//        Object value();

    @Target(ElementType.ANNOTATION_TYPE)
    @Retention(RetentionPolicy.RUNTIME)
    @interface List {
        Attribute[] value();
    }
}
