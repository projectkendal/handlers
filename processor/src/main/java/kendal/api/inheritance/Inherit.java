package kendal.api.inheritance;

import java.lang.annotation.*;

/**
 * 1. This annotation causes all parameters from #value() to be inherited.
 *    #value() must be an annotation.
 *
 * 2. Kendal handlers handling #value() will receive also annotations that inherit parameters from #value() using {@link Inherit}.
 * Note that this relation is transitive.
 */
@Target(ElementType.ANNOTATION_TYPE)
@Retention(RetentionPolicy.RUNTIME)
public @interface Inherit {
    //commented out, but we process it
//        Annotation value();


    /**
     * Metadata attribute present on {@link Inherit} annotation usages after compiling inheriting annotation.
     * Its usage in standard way should be avoided and will be discarded.
     *
     * Default is present here only to suppress errors highlighted by IDE, caused by lack of value passed.
     * It will be overriden anyway by Kendal.
     * @return
     */
    Class inheritedAnnotationClass() default Annotation.class;
}
