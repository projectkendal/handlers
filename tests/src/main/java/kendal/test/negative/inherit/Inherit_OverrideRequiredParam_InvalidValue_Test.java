package kendal.test.negative.inherit;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import kendal.api.inheritance.Inherit;

/*
 * @test
 * @summary check whether @Inherit does not work when trying override required parameter with value of wrong type
 * @build BaseAnn
 * @compile/fail/ref=Inherit_OverrideRequiredParam_InvalidValue_Test.out Inherit_OverrideRequiredParam_InvalidValue_Test.java -XDrawDiagnostics
 */
@SuppressWarnings("unused")
public class Inherit_OverrideRequiredParam_InvalidValue_Test {

    @Target(ElementType.METHOD)
    @Retention(RetentionPolicy.RUNTIME)
    @Inherit(@BaseAnn(requiredParam = "string"))
    @interface InheritingAnnOverridingParam{

    }

    @InheritingAnnOverridingParam
    private void anotherMethod() {

    }

}
