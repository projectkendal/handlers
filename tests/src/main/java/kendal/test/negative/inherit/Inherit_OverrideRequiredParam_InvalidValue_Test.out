Inherit_OverrideRequiredParam_InvalidValue_Test.java:21:39: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.String, int)
1 error